### PR TITLE
Add more lint and static analysis tools

### PR DESCRIPTION
**Pull request by weiiwang01**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -42,3 +42,11 @@ copyright-regexp = "Copyright\\s\\d{4}([-,]\\d{4})*\\s+%(author)s"
 markers = [
     "slow: marks slow and not very important tests"
 ]
+
+[tool.mypy]
+ignore_missing_imports = true
+allow_redefinition = true
+
+[tool.pylint.'MESSAGES CONTROL']
+max-line-length = 120
```
Comment (jdkandersson) [inline pyproject.toml:51]: Should this be 99/ 100?

File: src/charm.py [hunk 2]
```diff
@@ -794,9 +792,9 @@ def _addon_reconciliation(self, addon_type):
             addon_type (str): ``"theme"`` or ``"plugin"``.
         """
         self._check_addon_type(addon_type)
-        logger.debug(f"Start {addon_type} reconciliation process")
+        logger.debug("Start %s reconciliation process", addon_type)
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:795]: f-strings are a security risk if they are used in logging messages: https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#formatting-log-messages The team standard is to use f-strings except for log messages where the inbuilt message formatting should be used

File: tests/conftest.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
 
+# pylint: disable=missing-module-docstring
```
Comment (jdkandersson) [inline tests/conftest.py:4]: Alternatively, it could just be `"""Fixtures for tests"""`

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
 
+# pylint: disable=missing-module-docstring,missing-function-docstring
```
Comment (jdkandersson) [inline tests/integration/conftest.py:4]: For `missing-module-docstring`, it could just be `"""Fixtures for integration tests"""`

Comment (jdkandersson) [reply tests/integration/conftest.py:4]: I spotted 3 functions without a docstring, is it worth just adding a docstring for them?

File: tests/integration/test_core.py [hunk 5]
```diff
@@ -41,7 +48,7 @@ async def test_build_and_deploy(ops_test: pytest_operator.plugin.OpsTest, applic
     await ops_test.model.wait_for_idle()
     for unit in ops_test.model.applications[application_name].units:
         assert (
-            unit.workload_status == ops.model.BlockedStatus.name
+            unit.workload_status == ops.model.BlockedStatus.name  # type: ignore
```
Comment (jdkandersson) [inline tests/integration/test_core.py:51]: Is it worth including a comment with the reason for disabling?

**Review by jdkandersson — COMMENTED** [review]
Thanks for putting this together, there are some excellent improvements in here!

File: pyproject.toml (hunk 1 above)
Comment (weiiwang01) [reply pyproject.toml:51]: Fixed, thanks!

File: tests/conftest.py (hunk 3 above)
Comment (weiiwang01) [reply tests/conftest.py:4]: Added, thanks!

File: tests/integration/conftest.py (hunk 4 above)
Comment (weiiwang01) [reply tests/integration/conftest.py:4]: Fixed, thanks!

File: tests/integration/test_core.py (hunk 5 above)
Comment (weiiwang01) [reply tests/integration/test_core.py:51]: Fixed, thanks!

**Review by jdkandersson — APPROVED** [review]

**Review by arturo-seijas — COMMENTED** [review]
There are some linting errors

File: src/charm.py [hunk 6]
```diff
@@ -264,15 +266,15 @@ def _replica_consensus_reached(self):
             return False
         return all(replica_data.get(f) for f in fields)
 
-    def _on_leader_elected_replica_data_handler(self, event):
+    def _on_leader_elected_replica_data_handler(self, _event):
```
Comment (arturo-seijas) [inline src/charm.py:269]: Is there are reason why this is a named args instead of a throwaway arg`_`?

Comment (weiiwang01) [reply src/charm.py:269]: It's a convention to mark a unused argument.

Comment (weiiwang01) [reply src/charm.py:269]: I think both could work, I just prefer this way.

Comment (arturo-seijas) [reply src/charm.py:269]: I was just curious. Thx for the clarification

**Review by arturo-seijas — APPROVED** [review]
LGTM

**Pull request merged**